### PR TITLE
Reimplement "FreeAnchor: Learning to Match Anchors for Visual Object Detection"

### PR DESCRIPTION
**Pull request by yhcao6**
The code refers to https://github.com/zhangxiaosong18/FreeAnchor

| Model                     | Official | This repo |
|---------------------------|----------|-----------|
| FreeAnchor Retinanet R50  | 38.4     | 38.4      |
| FreeAnchor Retinanet R101 | 40.4     | 40.3      |

### CODE REVIEW

**Comment by mosheliv** [comment]
I am keen on trying this. How do I clone this into local system? I am not very familiar with github development cycle.

**Comment by yhcao6** [comment]
> I am keen on trying this. How do I clone this into local system? I am not very familiar with github development cycle.

`git clone -b free-anchor-ret https://github.com/yhcao6/mmdetection.git`

**Comment by mosheliv** [comment]
Thank you!

File: mmdet/models/anchor_heads/free_anchor_retina_head.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+import torch.nn.functional as F
+from torch.autograd import Function
+from torch.autograd.function import once_differentiable
+
+from mmdet.core import bbox2delta, bbox_overlaps, delta2bbox
+from ..registry import HEADS
+from .retina_head import RetinaHead
+
+
+class Clip(Function):
```
Comment (hellock) [inline mmdet/models/anchor_heads/free_anchor_retina_head.py]: Is it necessary to implement such a function?

**Comment by shuxp** [comment]
Is it possible to use FreeAnchor in two-stage detector?

File: mmdet/models/anchor_heads/free_anchor_retina_head.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        cls_prob = torch.sigmoid(cls_scores)
+        box_prob = []
+        num_pos = 0
+        positive_losses = []
+        for _, (anchors_, gt_labels_, gt_bboxes_, cls_prob_,
+                bbox_preds_) in enumerate(
+                    zip(anchors, gt_labels, gt_bboxes, cls_prob, bbox_preds)):
+            gt_labels_ -= 1
+
+            with torch.set_grad_enabled(False):
```
Comment (hellock) [inline mmdet/models/anchor_heads/free_anchor_retina_head.py]: `with torch.no_grad()`

File: mmdet/models/anchor_heads/free_anchor_retina_head.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                 bbox_threshold=0.6,
+                 gamma=2.0,
+                 alpha=0.5,
+                 **kwargs):
+        super(FreeAnchorRetinaHead,
+              self).__init__(num_classes, in_channels, stacked_convs,
+                             octave_base_scale, scales_per_octave, conv_cfg,
+                             norm_cfg, **kwargs)
+
+        self.iou_threshold = iou_threshold
```
Comment (hellock) [inline mmdet/models/anchor_heads/free_anchor_retina_head.py]: Better to follow our naming conventions like `iou_thr`.

**Pull request merged**